enzyme/Enzyme/MLIR/Analysis/DataFlowActivityAnalysis.cpp

```diff
@@ -0,0 +1,976 @@
+          auto *srcAliasClass =
+              getOrCreateFor<AliasClassLattice>(op, *copySource);
+          forEachAliasedAlloc(srcAliasClass, [&](DistinctAttr srcAlloc) {
+            if (before.hasActiveData(srcAlloc)) {
+              auto *destAliasClass =
+                  getOrCreateFor<AliasClassLattice>(op, value);
+              forEachAliasedAlloc(destAliasClass, [&](DistinctAttr destAlloc) {
+                result |= after->setActiveIn(destAlloc);
+              });
+            }
+          });
+        } else if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op)) {
+          // linalg.yield stores to the corresponding value.
+          for (OpOperand &dpsInit : linalgOp.getDpsInitsMutable()) {
+            if (dpsInit.get() == value) {
+              int64_t resultIndex =
+                  dpsInit.getOperandNumber() - linalgOp.getNumDpsInputs();
+              Value yieldOperand =
+                  linalgOp.getBlock()->getTerminator()->getOperand(resultIndex);
+              auto *valueState =
+                  getOrCreateFor<ForwardValueActivity>(op, yieldOperand);
+              if (valueState->getValue().isActiveVal()) {
+                auto *ptrAliasClass =
+                    getOrCreateFor<AliasClassLattice>(op, value);
+                forEachAliasedAlloc(ptrAliasClass, [&](DistinctAttr alloc) {
+                  result |= after->setActiveIn(alloc);
+                });
+              }
+            }
+          }
+        }
+      }
+    }
+    propagateIfChanged(after, result);
+  }
+
+  void visitCallControlFlowTransfer(CallOpInterface call,
+                                    CallControlFlowAction action,
+                                    const ForwardMemoryActivity &before,
+                                    ForwardMemoryActivity *after) override {
+    join(after, before);
+  }
+
+  /// Initialize the entry block with the supplied argument activities.
+  void setToEntryState(ForwardMemoryActivity *lattice) override {
+    if (auto *block = dyn_cast_if_present<Block *>(lattice->getPoint())) {
+      if (block == entryBlock) {
+        for (const auto &[arg, activity] :
+             llvm::zip(block->getArguments(), argumentActivity)) {
+          if (activity == enzyme::Activity::enzyme_dup ||
+              activity == enzyme::Activity::enzyme_dupnoneed) {
+            auto *argAliasClasses =
+                getOrCreateFor<AliasClassLattice>(block, arg);
+            for (DistinctAttr argAliasClass :
+                 argAliasClasses->getAliasClasses()) {
+              propagateIfChanged(lattice, lattice->setActiveIn(argAliasClass));
+            }
+          }
+        }
+      }
+    }
+  }
+
+private:
+  // A pointer to the entry block and argument activities of the top-level
+  // function being differentiated. This is used to set the entry state
+  // because we need access to the results of points-to analysis.
+  Block *entryBlock;
+  ArrayRef<enzyme::Activity> argumentActivity;
```

Comment:
(+) It's dangerous to store a non-owning reference here. I'd rather store a copy, the loss of space is marginal compared to the case where the analysis object outlive whoever owns this.